### PR TITLE
[SPARK-43026][SQL] Apply AQE with non-exchange table cache

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Apply AQE with non-exchange table cache at `InsertAdaptiveSparkPlan`

### Why are the changes needed?

`TableCacheQueryStageExec` supports to report runtime statistics, so it's possible that AQE  plans a better executed during re-optimization. Then it has benefits to apply AQE even without shuffle.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

add test

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala [hunk 1]
```diff
@@ -102,6 +103,8 @@ case class InsertAdaptiveSparkPlan(
         // update correctly.
         case i: InMemoryTableScanExec
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala:104]: shall we remove this case? If `CAN_CHANGE_CACHED_PLAN_OUTPUT_PARTITIONING` is false, `InMemoryTableScanExec` will not be AQEed.

Comment (ulysses-you) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/InsertAdaptiveSparkPlan.scala:104]: They can have different configurations, it is a cornor case though, 
1. people first tune `CAN_CHANGE_CACHED_PLAN_OUTPUT_PARTITIONING` on and cache a plan
2. people tune `CAN_CHANGE_CACHED_PLAN_OUTPUT_PARTITIONING` off and run that plan

then if we remove this pattern, then it won't apply AQE for the execued plan..

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Review by dongjoon-hyun — COMMENTED** [review]
+1, late LGTM. Thank you!

**Pull request closed without merging**